src/main/java/org/opensearch/ad/task/ADBatchTaskRunner.java

```diff
@@ -177,58 +197,395 @@ public ADBatchTaskRunner(
+            handleException(adTask, e);
+
+            if (adTask.isEntityTask()) {
+                // if (adTaskManager.isRetryableError(adTask.getError())
+                // && !adTaskCacheManager.exceedRetryLimit(adTask.getDetectorId(), adTask.getTaskId())) {
+                // // If the error is retryable, move entity from temp queue to the end of pending queue
+                // adTaskCacheManager.pushBackEntity(adTask.getTaskId(), adTask.getDetectorId(), adTask.getEntity().get(0).getValue());
+                // } else {
+                // adTaskCacheManager.removeEntity(adTask.getDetectorId(), adTask.getEntity().get(0).getValue());
+                // logger.warn("Entity task failed, task id: {}", adTask.getTaskId());
+                // }
+                // When reach this line, it means entity task failed to start on worker node
+                // Sleep some time before polling next entity task.
+                waitBeforeNextEntity(SLEEP_TIME_FOR_NEXT_ENTITY_TASK_IN_MILIS);
+                adTaskManager.entityTaskDone(adTask, e, transportService);
+                startNewEntityTaskLane(adTask, transportService);
+            }
+        });
+
+        ThreadedActionListener<ADBatchAnomalyResultResponse> threadedActionListener = new ThreadedActionListener<>(
+            logger,
+            threadPool,
+            AD_BATCH_TASK_THREAD_POOL_NAME,
+            actionListener,
+            false
+        );
+        return threadedActionListener;
+    }
+
+    private void waitBeforeNextEntity(long time) {
+        try {
+            Thread.sleep(time);
+        } catch (InterruptedException interruptedException) {
+            logger.warn("Exception while waiting", interruptedException);
+        }
+    }
+
+    private void forwardOrExecuteEntityTask(
+        ADTask adTask,
+        TransportService transportService,
+        ActionListener<ADBatchAnomalyResultResponse> delegatedListener
+    ) {
+        dispatchTask(adTask, ActionListener.wrap(node -> {
+            if (clusterService.localNode().getId().equals(node.getId())) {
+                // Execute batch task locally
+                startADBatchTaskOnWorkerNode(adTask, false, transportService, delegatedListener);
+            } else {
+                // Execute batch task remotely
+                transportService
+                    .sendRequest(
+                        node,
+                        ADBatchTaskRemoteExecutionAction.NAME,
+                        new ADBatchAnomalyResultRequest(adTask),
+                        option,
+                        // TODO: check if still need to add true/false in startADBatchTask
+                        new ActionListenerResponseHandler<>(delegatedListener, ADBatchAnomalyResultResponse::new)
+                    );
+            }
+        }, e -> delegatedListener.onFailure(e)));
+    }
+
+    // start new entity task lane
+    private void startNewEntityTaskLane(ADTask adTask, TransportService transportService) {
```

Comment:
Copy from https://github.com/opensearch-project/anomaly-detection/pull/58#discussion_r658364626

> Is one lane for each entity task? After finishing one entity task, you start a new lane by calling this function?

Entity task lane is a virtual concept. Will start a new lane before current entity tasks finish if not reach max running entities limit. If running entity done, it will start a new entity task in the same lane by sending back entity task done message to coordinating node.